code/game/objects/effects/decals/cleanable/humans.dm

```diff
@@ -1,194 +1,216 @@
 /obj/effect/decal/cleanable/blood
 	name = "blood"
-	desc = "It's red and gooey. Perhaps it's the chef's cooking?"
+	desc = "It's gooey. Perhaps it's the chef's cooking?"
 	icon = 'icons/effects/blood.dmi'
 	icon_state = "floor1"
 	random_icon_states = list("floor1", "floor2", "floor3", "floor4", "floor5", "floor6", "floor7")
-	blood_state = BLOOD_STATE_HUMAN
+	blood_state = BLOOD_STATE_BLOOD
+	color = BLOOD_COLOR_HUMAN //default so we don't have white splotches everywhere.
 	bloodiness = BLOOD_AMOUNT_PER_DECAL
 
+/obj/effect/decal/cleanable/blood/ComponentInitialize()
+	. = ..()
+	AddComponent(/datum/component/forensics)
+	update_icon()
+
 /obj/effect/decal/cleanable/blood/replace_decal(obj/effect/decal/cleanable/blood/C)
-	C.add_blood_DNA(return_blood_DNA())
-	if (bloodiness)
-		if (C.bloodiness < MAX_SHOE_BLOODINESS)
+	if(bloodiness)
+		if(C.bloodiness < MAX_SHOE_BLOODINESS)
 			C.bloodiness += bloodiness
+	update_icon()
 	return ..()
 
+/obj/effect/decal/cleanable/blood/add_blood_DNA(list/blood_dna)
+	return TRUE
+
+/obj/effect/decal/cleanable/blood/transfer_mob_blood_dna()
+	. = ..()
+	update_icon()
+
+/obj/effect/decal/cleanable/blood/update_icon()
+	GET_COMPONENT(F, /datum/component/forensics)
+	if(F && istype(F))
+		blood_color = F.blood_mix_color

```

Comment:
In case you don't know me I'm basically the primary datum component system maintainer on tg, greetings. I've been summoned here to try and save some components/signals from abuse.

You should not be using GetComponent except in rare cases where an object is basically owned by said component. Even then it is questionable. Take a look at components like squeak for a good example of how to isolate functionality.

You can replace the above code with code like the following:
```
/update_color()
  if(SEND_SIGNAL(COMSIG_UPDATE_COLOR) & COMPONENT_BLOCK_UPDATE_COLOR)
    return

[Inside component receiver of COMSIG_UPDATE_COLOR]
  parent.color = blood_mix_color
  return COMPONENT_BLOCK_UPDATE_COLOR
```

Components should be the holders of all logic related to their own feature. External code to forensics/whatever should not be in charge of how that component behaves. Do not use components as data holders.

If you ever need help with components or signals in any way feel free to pester me, using them requires a different way of going about problems and it can be difficult to pick up until it clicks.

Reply:
The benefits here by the way are many. You make it easier to maintain by making all code related to a functionality contained in the object controlling that functionality. You make it easier to extend or modify that functionality for the same reason. You avoid bugs by not caring what the component is applied to or what's its parent is doing. You even get some significant performance benefits by not using the fairly costly GetComponent call or checking if the component exists after.

As a side note you never need to istype the return value.

Reply:
I'll blame Bhjin for suggesting the istype bit.

As I said, I very much appreciate this write up, even if I'm still a bit too dense to really understand it. most of the component bad-touch was my growing increasingly desperate in a system that just doesn't have enough comparable implementations to what I'd want at an easy at-a-glance sort of learning.

Reply:
I'd suggest you take a look at the more self contained components like, thermite, squeak, and waddle for examples